In [1]:
import numpy as np
import pandas as pd
import re
from tqdm import tqdm
import datetime as datetime

# BPI Challenge 2012

In [50]:
df_12 = pd.read_csv("BPI_Challenge_2012.csv", parse_dates = ['time:timestamp'])

In [11]:
df_12.describe(include='all')

Unnamed: 0   org:resource lifecycle:transition  \
count   262200.000000  244190.000000               262200   
unique            NaN            NaN                    3   
top               NaN            NaN             COMPLETE   
freq              NaN            NaN               164506   
mean    131099.500000    8943.364687                  NaN   
std      75690.764628    4242.517645                  NaN   
min          0.000000     112.000000                  NaN   
25%      65549.750000   10609.000000                  NaN   
50%     131099.500000   10931.000000                  NaN   
75%     196649.250000   11122.000000                  NaN   
max     262199.000000   11339.000000                  NaN   

                  concept:name                    time:timestamp  \
count                   262200                            262200   
unique                      24                            248189   
top     W_Completeren aanvraag  2012-01-05 10:01:25.057000+01:00   
freq                     54850                                 4   
mean                       NaN                               NaN   
std                        NaN                               NaN   
min                        NaN                               NaN   
25%                        NaN                               NaN   
50%                        NaN                               NaN   
75%                        NaN                               NaN   
max                        NaN                               NaN   

                           case:REG_DATE  case:concept:name  case:AMOUNT_REQ  
count                             262200      262200.000000    262200.000000  
unique                             13087                NaN              NaN  
top     2011-11-15 13:42:45.592000+01:00                NaN              NaN  
freq                                 175                NaN              NaN  
mean                                 NaN      193811.496754     15586.795381  
std                                  NaN       11668.794018     12381.430915  
min                                  NaN      173688.000000         0.000000  
25%                                  NaN      183277.000000      6500.000000  
50%                                  NaN      194206.000000     11000.000000  
75%                                  NaN      203802.000000     20000.000000  
max                                  NaN      214376.000000     99999.000000

## 2012 dataset attributes

**org:resource** -  integer or NaN (identifies some instance)

**lifecycle:transition** - status of request (COMPLETE/START/SCHEDULE)

**concept:name** - type of request (?)

**time:timestamp** 

**case:REG_DATE** - registration date (== first timestamp)

**case:concept:name** - integer - identifies the case?

**case:AMOUNT_REQ** - loan amount applied for (single for the whole case)

# BPI Challenge 2017

In [49]:
df_2017 = pd.read_csv("BPI Challenge 2017.csv", parse_dates = ['time:timestamp'])

In [31]:
df_2017.describe(include='all')

/var/folders/lh/s2l49w850tq2lvbfyqst82j00000gn/T/ipykernel_4265/1223815370.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  df_2017.describe(include='all')


Unnamed: 0       Action org:resource            concept:name  \
count   1.202267e+06      1202267      1202267                 1202267   
unique           NaN            5          149                      26   
top              NaN  statechange       User_1  W_Validate application   
freq             NaN       358940       148404                  209496   
first            NaN          NaN          NaN                     NaN   
last             NaN          NaN          NaN                     NaN   
mean    6.011330e+05          NaN          NaN                     NaN   
std     3.470647e+05          NaN          NaN                     NaN   
min     0.000000e+00          NaN          NaN                     NaN   
25%     3.005665e+05          NaN          NaN                     NaN   
50%     6.011330e+05          NaN          NaN                     NaN   
75%     9.016995e+05          NaN          NaN                     NaN   
max     1.202266e+06          NaN          NaN                     NaN   

       EventOrigin                EventID lifecycle:transition  \
count      1202267                1202267              1202267   
unique           3                1202267                    7   
top       Workflow  Application_652823628             complete   
freq        768823                      1               475306   
first          NaN                    NaN                  NaN   
last           NaN                    NaN                  NaN   
mean           NaN                    NaN                  NaN   
std            NaN                    NaN                  NaN   
min            NaN                    NaN                  NaN   
25%            NaN                    NaN                  NaN   
50%            NaN                    NaN                  NaN   
75%            NaN                    NaN                  NaN   
max            NaN                    NaN                  NaN   

                          time:timestamp case:LoanGoal case:ApplicationType  \
count                            1202267       1202267              1202267   
unique                           1202209            14                    2   
top     2016-07-04 13:54:42.993000+00:00           Car           New credit   
freq                                   2        339798              1083767   
first   2016-01-01 09:51:15.304000+00:00           NaN                  NaN   
last    2017-02-01 14:11:03.499000+00:00           NaN                  NaN   
mean                                 NaN           NaN                  NaN   
std                                  NaN           NaN                  NaN   
min                                  NaN           NaN                  NaN   
25%                                  NaN           NaN                  NaN   
50%                                  NaN           NaN                  NaN   
75%                                  NaN           NaN                  NaN   
max                                  NaN           NaN                  NaN   

             case:concept:name  case:RequestedAmount  FirstWithdrawalAmount  \
count                  1202267          1.202267e+06           42995.000000   
unique                   31509                   NaN                    NaN   
top     Application_2037628374                   NaN                    NaN   
freq                       180                   NaN                    NaN   
first                      NaN                   NaN                    NaN   
last                       NaN                   NaN                    NaN   
mean                       NaN          1.675947e+04            8394.338979   
std                        NaN          1.572320e+04           10852.443358   
min                        NaN          0.000000e+00               0.000000   
25%                        NaN          6.000000e+03               0.000000   
50%                        NaN          1.400000e+04            5000.000

In [46]:
df_2017['Accepted'].unique()

array([nan, True, False], dtype=object)

## 2017 dataset attributes

1. **Action** - 'Created', 'statechange', 'Deleted', 'Obtained', 'Released'

2. **org:resource** -  149 unique User names

3. **concept:name** - 'A_Create Application', 'A_Submitted', 'W_Handle leads', 'W_Complete application', 'A_Concept', 'A_Accepted', 'O_Create Offer', 'O_Created', 'O_Sent (mail and online)', 'W_Call after offers', 'A_Complete', 'W_Validate application', 'A_Validating', 'O_Returned', 'W_Call incomplete files', 'A_Incomplete', 'O_Accepted', 'A_Pending', 'A_Denied', 'O_Refused', 'O_Cancelled', 'A_Cancelled', 'O_Sent (online only)', 'W_Assess potential fraud', 'W_Personal Loan collection', 'W_Shortened completion '  --> ***much more clear than in 2012 dataset!!***
  
4. **EventOrigin** - 'Application', 'Workflow', 'Offer'

5. **EventID** - format:Application_652823628 - every entry unique - key

6. **lifecycle:transition** - 'complete', 'schedule', 'withdraw', 'start', 'suspend', 'ate_abort', 'resume' --> ***2017 has 7 states, 2012 only 3***
       
7. **time:timestamp**

8. **case:LoanGoal** - what is the loan for (14 unique values)

9. **case:ApplicationType** - 'New credit', 'Limit raise'

10. **case:concept:name** - (probably) identifies whole individual cases

11. **case:RequestedAmount**

12. **FirstWithdrawalAmount** - only where request accepted - else, NaN

13. **NumberOfTerms** - only where request accepted

14. **Accepted** - bool (***more straightforward than in 2012***) - but can be nan

15. **MonthlyCost** - only where request accepted

16. **Selected** - only where request accepted

17. **CreditScore** - only where request accepted

18. **OfferedAmount** - only where request accepted

19. **OfferID** - identifies an offer, there are as many unique vals as there are accepted requests

# BPI Challenge 2018

In [51]:
df_2018 = pd.read_csv("BPI Challenge 2018.csv", parse_dates = ['time:timestamp'])

In [47]:
df_2018.describe(include='all')

Unnamed: 0  success org:resource  \
count   2.514266e+06  2514266      2514266   
unique           NaN        2          165   
top              NaN     True       727350   
freq             NaN  2481899       748950   
mean    1.257132e+06      NaN          NaN   
std     7.258062e+05      NaN          NaN   
min     0.000000e+00      NaN          NaN   
25%     6.285662e+05      NaN          NaN   
50%     1.257132e+06      NaN          NaN   
75%     1.885699e+06      NaN          NaN   
max     2.514265e+06      NaN          NaN   

                                  docid_uuid              doctype  \
count                                2514266              2514266   
unique                                225270                    8   
top     8A28496E-4393-41BA-96BA-FC6F353F1028  Payment application   
freq                                     780               984613   
mean                                     NaN                  NaN   
std                                      NaN                  NaN   
min                                      NaN                  NaN   
25%                                      NaN                  NaN   
50%                                      NaN                  NaN   
75%                                      NaN                  NaN   
max                                      NaN                  NaN   

         subprocess         docid   activity     note       eventid  ...  \
count       2514266  2.514266e+06    2514266  2508880  2.396132e+06  ...   
unique            8           NaN         41      137           NaN  ...   
top     Application           NaN  calculate     none           NaN  ...   
freq         928091           NaN     466141  1233571           NaN  ...   
mean            NaN -3.734307e+16        NaN      NaN -5.646112e+16  ...   
std             NaN  2.664162e+16        NaN      NaN  4.058381e+16  ...   
min             NaN -1.441092e+17        NaN      NaN -1.621238e+17  ...   
25%             NaN -5.403729e+16        NaN      NaN -9.006608e+16  ...   
50%             NaN -1.800862e+16        NaN      NaN -5.403716e+16  ...   
75%             NaN -1.800852e+16        NaN      NaN -1.800853e+16  ...   
max             NaN  1.801499e+17        NaN      NaN  3.602939e+17  ...   

       case:concept:name case:penalty_amount1 case:payment_actual1  \
count            2514266        262017.000000        262017.000000   
unique             43809                  NaN                  NaN   
top     93deb55be9aed2be                  NaN                  NaN   
freq                2973                  NaN                  NaN   
mean                 NaN           732.043284           137.124164   
std                  NaN          1984.368225          2299.169823   
min                  NaN             0.000000        -11945.800000   
25%                  NaN            70.500000          -245.850000   
50%                  NaN           187.390000          -121.820000   
75%                  NaN           541.290000             0.000000   
max                  NaN         31241.380000         35716.710000   

       case:amount_applied1 case:penalty_amount2 case:payment_actual2  \
count         262017.000000         22550.000000         22550.000000   
unique                  NaN                  NaN                  NaN   
top                     NaN                  NaN                  NaN   
freq                    NaN                  NaN                  NaN   
mean           67727.301910          1214.824002           -87.904554   
std           118248.981412          2802.493763          2299.845439   
min                0.000000             0.000000        -17421.930000   
25%            19262.490000           194.770000          -364.810000   
50%            36134.210000           450.840000             0.000000   
75%            62578.540000          1008.320000             0.000000   
max           608980.200000         31241.390000         18186.340000  

## 2018 dataset attributes

In general, the structre is similar but there are much more detailed attributes about each case. There are more identifiers (but hard to tell at first glance what they are for). There is the distinctions for amount applied, penalties, payments - they indicate if someone paid the whole expected sum, if there was a penalty (probably for a delay in payment) etc.

# Road Traffic Fine Management Process

In [7]:
df_traffic = pd.read_csv("Road_Traffic_Fine_Management_Process.csv", parse_dates = ['time:timestamp'])

df_traffic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 561470 entries, 0 to 561469
Data columns (total 17 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Unnamed: 0            561470 non-null  int64  
 1   amount                230230 non-null  float64
 2   org:resource          150925 non-null  float64
 3   dismissal             155066 non-null  object 
 4   concept:name          561470 non-null  object 
 5   vehicleClass          150370 non-null  object 
 6   totalPaymentAmount    227971 non-null  float64
 7   lifecycle:transition  561470 non-null  object 
 8   time:timestamp        561470 non-null  object 
 9   article               150370 non-null  float64
 10  points                150370 non-null  float64
 11  case:concept:name     561470 non-null  object 
 12  expense               103987 non-null  float64
 13  notificationType      79860 non-null   object 
 14  lastSent              78229 non-null   object 
 15  

In [14]:
df_traffic.describe(include='all')

Unnamed: 0         amount   org:resource dismissal concept:name  \
count   561470.000000  230230.000000  150925.000000    155066       561470   
unique            NaN            NaN            NaN        26           11   
top               NaN            NaN            NaN       NIL  Create Fine   
freq              NaN            NaN            NaN    152484       150370   
mean    280734.500000      63.681656     443.611383       NaN          NaN   
std     162082.572158      89.677991     300.505385       NaN          NaN   
min          0.000000       0.000000       0.000000       NaN          NaN   
25%     140367.250000      33.600000      50.000000       NaN          NaN   
50%     280734.500000      38.000000     541.000000       NaN          NaN   
75%     421101.750000      71.500000     561.000000       NaN          NaN   
max     561469.000000    8000.000000     870.000000       NaN          NaN   

       vehicleClass  totalPaymentAmount lifecycle:transition  \
count        150370       227971.000000               561470   
unique            4                 NaN                    1   
top               A                 NaN             complete   
freq         146712                 NaN               561470   
mean            NaN           17.815604                  NaN   
std             NaN           36.779382                  NaN   
min             NaN            0.000000                  NaN   
25%             NaN            0.000000                  NaN   
50%             NaN            0.000000                  NaN   
75%             NaN           35.000000                  NaN   
max             NaN         4021.000000                  NaN   

                   time:timestamp        article         points  \
count                      561470  150370.000000  150370.000000   
unique                       4905            NaN            NaN   
top     2002-04-10 00:00:00+02:00            NaN            NaN   
freq                         9197            NaN            NaN   
mean                          NaN     113.246658       0.078879   
std                           NaN      68.219475       0.577687   
min                           NaN       7.000000       0.000000   
25%                           NaN       7.000000       0.000000   
50%                           NaN     157.000000       0.000000   
75%                           NaN     157.000000       0.000000   
max                           NaN     401.000000      10.000000   

       case:concept:name        expense notificationType lastSent  \
count             561470  103987.000000            79860    78229   
unique            150370            NaN                2        3   
top               C20817            NaN                P        P   
freq                  20            NaN            79726    46476   
mean                 NaN      11.948984              NaN      NaN   
std                  NaN       4.023190              NaN      NaN   
min                  NaN       0.000000              NaN      NaN   
25%                  NaN      10.000000              NaN      NaN   
50%                  NaN      11.880000              NaN      NaN   
75%                  NaN      13.500000              NaN      NaN   
max                  NaN      76.000000              NaN      NaN   

        paymentAmount  matricola  
count    77601.000000      555.0  
unique            NaN        NaN  
top               NaN        NaN  
freq              NaN        NaN  
mean        46.410027        0.0  
std         39.345561        0.0  
min          0.000000        0.0  
25%         33.600000        0.0  
50%         36.000000        0.0  
75%         44.670000        0.0  
max       3975.000000        0.0

In [23]:
df_traffic['matricola'].unique()

array([nan,  0.])

## Italian road

**org:resource** -  integer or NaN (identifies some instance)

**dismissal** - NIL, nan, #, G, N, K, 5, '3', 'A', 'I', 'D', 'T', 'E', '@', 'M', 'Q', 'F', 'V', 'U', 'C', 'B', '$', 'Z', 'J', 'R', '2', '4'

**concept:name** - 'Create Fine', 'Send Fine', 'Insert Fine Notification', 'Add penalty', 'Send for Credit Collection', 'Payment', 'Insert Date Appeal to Prefecture', 'Send Appeal to Prefecture', 'Receive Result Appeal from Prefecture', 'Notify Result Appeal to Offender', 'Appeal to Judge'

**vehicleClass** - 'A', nan, 'C', 'M', 'R'

**totalPaymentAmount** - fine amount - most are 0

**lifecycle:transition** - all are "complete"

**time:timestamp** - 4905 unique - number of cases

**article** - integer - probably signifies the type of offense

**points** - loan ammount applied for (single for the whole case)

**case:concept:name** - identifies case stages????

**expense**

**notificationType** - nan, 'P', 'C'

**lastSent** - nan, 'P', 'N', 'C'

**paymentAmount** - how is it different from totalPaymentAmount??

**matricola** - nan or 0. - probably irrelevant

--> in general, the values are less intuitive and seem less structured than in the BPI datasets